### PR TITLE
Add App Startup library

### PR DESCRIPTION
**Pull request by dev-weiqi**
First of all thank you for this awesome project.
I added [App Startup](https://developer.android.com/topic/libraries/app-startup) library in this project and the user doesn't need to initialize manually anymore!

I also fixed the build issues.  ActivityLifecycleCallbackExtTest using reflect modifiers. It's no longer supporting on JDK16 or later.
(I'm not native speaker, sorry for any grammatical mistakes)

- [x] ./gradlew build -> BUILD SUCCESSFUL 
- [x] ./gradlew spotlessApply -> BUILD SUCCESSFUL

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -49,23 +49,9 @@ dependencies {
 
 _You can find latest version and changelogs in the [releases](https://github.com/PatilShreyas/permission-flow-android/releases)_.
 
-### 2. Initialization
+### 2. Observing a Permission State
 
-Before using any API of the library, it needs to be initialized before.
```
Comment (PatilShreyas) [inline README.md:54]: Do not remove this section. Instead, we should explain that this library automatically gets initialized with the App Startup library. So if someone disables the initializer (_to provide their own coroutine dispatcher_), there should be a proper resource for them to refer to.

File: permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/PermissionFlow.kt [hunk 2]
```diff
@@ -180,7 +180,7 @@ interface PermissionFlow {
      * getting instance.
      */
     companion object {
-        private val DEFAULT_DISPATCHER = Executors.newFixedThreadPool(2).asCoroutineDispatcher()
+        internal val DEFAULT_DISPATCHER = Executors.newFixedThreadPool(2).asCoroutineDispatcher()
```
Comment (PatilShreyas) [inline permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/PermissionFlow.kt:183]: Keep it private. Do not change this

File: permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/PermissionFlow.kt [hunk 3]
```diff
@@ -189,8 +189,7 @@ interface PermissionFlow {
          * @param dispatcher Coroutine dispatcher to be used in the [PermissionFlow]. By default,
          * it utilizes dispatcher having fixed two number of threads.
          */
-        @JvmStatic
-        @JvmOverloads
+        @Deprecated(message = "We don't need to call `init()` in this version.", level = DeprecationLevel.HIDDEN)
```
Comment (PatilShreyas) [inline permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/PermissionFlow.kt:192]: This method shouldn't be marked as deprecated, this is still needed to be able to configure the dispatcher.

File: permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/initializer/PermissionFlowInitializer.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import dev.shreyaspatil.permissionFlow.PermissionFlow
+import dev.shreyaspatil.permissionFlow.impl.PermissionFlowImpl
+
+/**
+ * Initializes [PermissionFlow] instance on app startup.
+ */
+class PermissionFlowInitializer : Initializer<Unit> {
+
+    override fun create(context: Context) {
+        PermissionFlowImpl.init(context, PermissionFlow.DEFAULT_DISPATCHER)
```
Comment (PatilShreyas) [inline permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/initializer/PermissionFlowInitializer.kt:29]: This method is overloaded already, no need to provide a default dispatcher already. 

```suggestion
        PermissionFlowImpl.init(context)
```

File: permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/utils/ApiLevelChecker.kt [hunk 5]
```diff
@@ -13,14 +13,17 @@
...
-import android.app.Application
-import dev.shreyaspatil.permissionFlow.PermissionFlow
+import android.annotation.SuppressLint
+import android.os.Build
 
-class ExampleApplication : Application() {
-    override fun onCreate() {
-        super.onCreate()
-        PermissionFlow.init(this)
+internal object ApiLevelChecker {
```
Comment (PatilShreyas) [inline permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/utils/ApiLevelChecker.kt:21]: This is nice addition 👍🏻

**Review by PatilShreyas — CHANGES_REQUESTED** [review]
@dev-weiqi Thanks for the PR, it's really nice 😃. Here are some comments.

**Comment by PatilShreyas** [comment]
@dev-weiqi CodeCov is failing due to coverage impact, could you please add basic tests for the initializer?

File: README.md (hunk 1 above)
Comment (dev-weiqi) [reply README.md:54]: Done. 
Thanks for the suggestion, learn a lot😀

File: permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/PermissionFlow.kt (hunk 2 above)
Comment (dev-weiqi) [reply permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/PermissionFlow.kt]: Done.

File: permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/initializer/PermissionFlowInitializer.kt (hunk 4 above)
Comment (dev-weiqi) [reply permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/initializer/PermissionFlowInitializer.kt]: Done.

File: permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/utils/ApiLevelChecker.kt (hunk 5 above)
Comment (dev-weiqi) [reply permission-flow/src/main/java/dev/shreyaspatil/permissionFlow/utils/ApiLevelChecker.kt]: I was using the wrong JDK version 😱 , I think we don't need this class anymore.

File: README.md (hunk 1 above)
Comment (PatilShreyas) [reply README.md:54]: Nice

File: README.md [hunk 6]
```diff
@@ -51,18 +51,38 @@ _You can find latest version and changelogs in the [releases](https://github.com
 
 ### 2. Initialization
 
-Before using any API of the library, it needs to be initialized before.   
-Initialize **PermissionFlow** as follows (For example, in `Application` class):
+This library automatically gets initialized with the App Startup library.
+If you want to provide own coroutine dispatcher
```
Comment (PatilShreyas) [inline README.md:55]: Don't say "_If you want to provide own coroutine dispatcher_". Instead say, "If you want to initialize manually". Also suggest them to remove the initializer provider from the Manifest.

```xml
<provider
    android:name="androidx.startup.InitializationProvider"
    android:authorities="${applicationId}.androidx-startup"
    android:exported="false"
    tools:node="merge">
    
    <meta-data
        android:name="dev.shreyaspatil.permissionFlow.initializer.PermissionFlowInitializer"
        android:value="androidx.startup"
        tools:node="remove" />
 </provider>
```

Remove the below section of "Disable PermissionFlowInitializer in AndroidManifest.xml" and keep it here only

File: README.md [hunk 7]
```diff
@@ -51,18 +51,38 @@ _You can find latest version and changelogs in the [releases](https://github.com
...
+
+#### 2.1 Initialize **PermissionFlow** as follows (For example, in `Application` class)
 
 ```kotlin
 class MyApplication: Application() {
     override fun onCreate() {
         super.onCreate()
-        PermissionFlow.init(this)
+        val permissionDispatcher = Executors.newFixedThreadPool(3).asCoroutineDispatcher()
+        PermissionFlow.init(this, permissionDispatcher)
```
Comment (PatilShreyas) [inline README.md:64]: Just provide `Dispatchers.IO` in the example. Keep it simple and small as possible

File: README.md [hunk 8]
```diff
@@ -120,25 +140,25 @@ fun ExampleSinglePermission() {
 fun ExampleMultiplePermission() {
     val state by rememberMultiplePermissionState(
         Manifest.permission.CAMERA
-        Manifest.permission.ACCESS_FINE_LOCATION,
+                Manifest.permission.ACCESS_FINE_LOCATION,
```
Comment (PatilShreyas) [inline README.md:143]: Fix indentations

File: permission-flow/src/test/java/dev/shreyaspatil/permissionFlow/initializer/PermissionFlowInitializerTest.kt [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        verify(exactly = 1) { PermissionFlowImpl.init(context, any()) }
+    }
+
+    @Test
+    fun testInitializerDependencies_shouldBeEmpty() {
+        val initializer = PermissionFlowInitializer()
+
+        Assert.assertTrue(initializer.dependencies().isEmpty())
+    }
+}
```
Comment (PatilShreyas) [inline permission-flow/src/test/java/dev/shreyaspatil/permissionFlow/initializer/PermissionFlowInitializerTest.kt:49]: Good work here 👍🏻

**Review by PatilShreyas — CHANGES_REQUESTED** [review]

**Pull request merged**